File: src/go/1-25/problem_4.go
/*
                           Largest palindrome product

   A  palindromic  number  reads the same both ways. The largest palindrome made
   from the product of two 2-digit numbers is 9009 = 91 × 99.

   Find the largest palindrome made from the product of two 3-digit numbers.

                               Answer: 906609
*/

package main

import (
	"fmt"
	"math"
	"strconv"
)

func isPalindrome(number int) bool {
	read := strconv.Itoa(number)
	lenght := len(read)
	flag := true

	for i := 0; i < lenght/2; i++ {
		if read[i] != read[lenght-i-1] {
			flag = false
			break
		}
	}

	return flag
}

func largestPalindrome(digits float64, channel chan int) {
	var mult, max int

	for i := 1; i < int(math.Pow(10, digits)); i++ {
		for j := 1; j < int(math.Pow(10, digits)); j++ {
			mult = i * j

			if isPalindrome(mult) && mult > max {
				max = mult
			}
		}
	}

	channel <- max
}

func main() {
	var maxDigits float64 = 3
	channel := make(chan int, 1)

	go largestPalindrome(maxDigits, channel)

	fmt.Println(<-channel)
}

File: src/haskell/1-25/problem_4.hs
--                         Largest palindrome product

-- A  palindromic  number  reads the same both ways. The largest palindrome made
-- from the product of two 2-digit numbers is 9009 = 91 × 99.

-- Find the largest palindrome made from the product of two 3-digit numbers.

--                             Answer: 906609

isPalindrome :: (Eq a) => [a] -> Bool
isPalindrome [] = True
isPalindrome [_] = True
isPalindrome (x:xs) = if (==) x (last xs) then (isPalindrome . init) xs else False

productOfNDigits :: Integer -> [String]
productOfNDigits number = [ show(x*y) | x <- [lower..upper], y <- [x..upper] ] where
    lower = 10^((-) number 1) :: Integer
    upper = (-) (10^number) 1 :: Integer

main :: IO()
main = do
    let digits = 3 :: Integer
    print $ maximum $ map (read::String->Int) $ filter isPalindrome $ productOfNDigits digits
